atomic_reactor/utils/resource_manager.py

```diff
@@ -0,0 +1,187 @@
+                        ret = self._write_free_slot(slot.get('sid'))
+                        if ret == b'':
+                            logger.info('Unlocked')
+                            return True
+                        else:
+                            logger.debug('Unlocking failed, retrying')
+                            retries += 1
+        return False
+
+    def lock(self, slot: int, prid: str):
+        logger.debug("Aquiring lock")
+        payload = json.dumps({"prid": prid, "locked": str(datetime.now())})
+        retries = 0
+        while retries < 3:
+            filename = os.path.join(ROOT_PATH, f'slot_{slot}.json')
+            _cmd = 'if grep -qw 0 {0} ; then echo {1} > {0} ;else exit 1; fi'.format(
+                    filename,
+                    quote(payload)
+                    )
+            cmd = 'flock {} -c {}; echo $?'.format(filename, quote(_cmd))
+            logger.debug(cmd)
+            ret = self._ssh_run_remote_command(cmd)
+            if ret == b'0\n':
+                logger.info('Locked')
+                return True
+            else:
+                logger.debug('Locking failed, retrying')
+                retries += 1
+        logger.info('Locking failed')
+        return False
+
+    def _ssh_run_remote_command(self, cmd: str):
+        ssh_client = paramiko.SSHClient()
+        ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh_client.connect(hostname=self.hostname,
+                           username=self.username)
+        cmd_out = ssh_client.exec_command(cmd)
+        out = cmd_out[1].read()
+        error = cmd_out[2].read()
+        if error:
+            raise Exception('There was an error pulling the runtime: {}'.format(error))
+        ssh_client.close()
+        return out
+
+
+class ResourceManager:
+    '''
+    Manages build resources
+    '''
+
+    def __init__(self, hosts: list[Host]):
+        self.hosts = hosts
+
+    def find(self, arch: str) -> list[Host]:
+        try:
+            resources = []
+            for host in self.hosts:
+                if host.arch == arch:
+                    host.check_avail()
+                    if host.free_slots > 0:
+                        resources.append(host)
+            sorted_result = sorted(resources, key=lambda h: h.free_slots)
+            logger.debug('sorted result: %s', sorted_result)
+            return sorted_result
+        except Exception as err:
+            logger.debug(err)
+            logger.error("Unable to read hosts file")
+            return []
+
+    def obtain_free_resource(self, arch: str, prid: str) -> Host:
+        availiable_hosts = self.find(arch)
+        if availiable_hosts:
+            for host in availiable_hosts:
+                for slot in host.slots:
+                    if slot.get('state', '') == 'free':
+                        if host.lock(slot.get('sid', ''), prid):
+                            return host
+        return False
```

Comment:
I believe this method should return list of available hosts or empty list, not False.

Reply:
Fixed in last commit